CI/run_tests.sh

```diff
@@ -1,23 +1,32 @@
 #!/usr/bin/env bash
 
-sudo python ptf_nn/ptf_nn_agent.py \
+set -x
+
+MAJOR_MINOR_VERSION=$(python3 -c 'import sys; print(f"{sys.version_info.major}.{sys.version_info.minor}")')
+PPATH="$HOME/.local/lib/python${MAJOR_MINOR_VERSION}/site-packages"
+
+sudo PATH=${PATH} PYTHONPATH=${PPATH} python3 ptf_nn/ptf_nn_agent.py \
```

Comment:
Why are these path hacks necessary btw? 

Maybe we should consider setting up virtualenv or poetry if Ubuntu24.04 requires it.

Reply:
Ah, I guess because of the use of `sudo`. 

Although `sudo -E env PATH="$PATH"` should be enough here. If the Python environment is set up correctly.

Reply:
If you set up virtualenv as the normal user, then you pick a directory where that is installed, and using it typically either requires, or recommends, first running `source $MYVENV/bin/activate`, which might do other things, but perhaps the most crucial is setting the environment variable `VIRTUAL_ENV`.

If you know a way to run commands as root that don't do one of (a) pass the environment variable values on to the root user, as shown in my current sudo commands, or (b) somehow do `source $MYVENV/bin/activate ; <other command I really want to run as root>`, I'm interested to learn what that is.

I don't know of any way in which virtual environments somehow avoid this.

Reply:
And if I recall correctly, doing what I am doing now, where the root user uses a normal user's lib/python* directory, isn't merely read-only, but the root user can also write `.pyc` files and `__pycache__` directories that the normal user then does not have permission to modify later.

Reply:
I don't think there is a way around this unless you disable sudo's path protection: https://unix.stackexchange.com/a/151188

I have found `sudo -E env PATH="$PATH"` the least invasive approach

Reply:
Maybe `--preserve-env=PATH` as parameter works?

Reply:
And don't blame Ubuntu for this.  I think it is Python pip and some particular new version of it released within the last 2-4 years that makes you jump through hoops to install packages in system-wide directories.

Reply:
I think for the programs using sudo in these tests, at least, the main reason sudo is being used is that it is sending packets to a veth, or sniffing packets on a veth.  An alternative for that would be to somehow set up capabilities for the normal user so that they have permission to do this, and then run the tests as the normal (not root) user.

Reply:
Sorry, was looking things up and writing replies without reading your replies.  I will try `sudo -E env PATH="$PATH"`

Reply:
commit 25 uses `sudo -E env PATH="$PATH" ...` with success.  Thanks for the suggestion.

Reply:
https://unix.stackexchange.com/a/480764 

Might even remove the need to use any dollar signs.

Reply:
Trying that variation with commit 26 to see if it passes the tests.

At some point, we have to stop tweaking and use something that works :-)

Reply:
I was mostly interested because of https://github.com/p4lang/p4c/pull/5047
Some of the lessons here I would like to apply to that PR.